src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java

```diff
@@ -451,9 +464,19 @@ public void doInsertJson(List<Record> records) throws IOException, ExecutionExce
                 // write bytes into the piped stream
                 for (Record record: records ) {
                     if (record.getSinkRecord().value() != null ) {
-                        Map<String, Object> data = (Map<String, Object>) record.getSinkRecord().value();
-                        java.lang.reflect.Type gsonType = new TypeToken<HashMap>() {
-                        }.getType();
+                        Map<String, Object> data;
+                        if (record.getSinkRecord().value() instanceof Map) {
+                            data = (Map<String, Object>) record.getSinkRecord().value();
+                        } else if (record.getSinkRecord().value() instanceof Struct) {
+                            data = new HashMap<>();
+                            Struct struct = (Struct) record.getSinkRecord().value();
+                            for (Field field : struct.schema().fields()) {
+                                data.put(field.name(), struct.get(field));//Doesn't handle multi-level object depth
+                            }
+                        } else {
+                            throw new RuntimeException("Unsupported record type: " + record.getSinkRecord().value().getClass().getName());
+                        }
+                        java.lang.reflect.Type gsonType = new TypeToken<HashMap>() {}.getType();
```

Comment:
Maybe we can use data.clear() to be less aggressive on the GC so we do not need to create data = new HashMap<>();
Member